src/wp-includes/class-wp-query.php

```diff
@@ -3571,7 +3571,31 @@ public function the_post() {
 		global $post;
 
 		if ( ! $this->in_the_loop ) {
-			update_post_author_caches( $this->posts );
+			/*
+			 * Warm the author cache with full WP_Post objects.
+			 *
+			 * This prevents an error within `update_post_author_caches()` when
+			 * WP_Query is used to return a sub-set of fields.
+			 */
+			$post_ids = array_map(
+				function( $post ) {
+					if ( is_numeric( $post ) ) {
+						return $post;
+					}
+
+					if ( isset( $post->ID ) ) {
+						return $post->ID;
+					}
+
+					return 0;
+				},
+				$this->posts
+			);
+			$post_ids = array_filter( $post_ids );
+			_prime_post_caches( $post_ids, $this->query_vars['update_post_term_cache'], $this->query_vars['update_post_meta_cache'] );
+
+			$post_objects = array_map( 'get_post', $post_ids );
+			update_post_author_caches( $post_objects );
```

Comment:
```suggestion
			$post_ids = array_filter( $this->posts, 'is_numeric' );
			$post_ids = array_filter( $post_ids );
			if( $post_ids ){
			     _prime_post_caches( $post_ids, $this->query_vars['update_post_term_cache'], $this->query_vars['update_post_meta_cache'] );
                         }
			$post_objects = array_map( 'get_post', $this->posts );
			update_post_author_caches( $post_objects );
```

Something a little simpler.

Reply:
I don't think these do the same thing. In the first version a `WP_Post` object will be convered to it's id. In this version, it will be discarded by the `array_filter( 'is_numeric' )`

Reply:
When would there be a case where there would be a mixture of WP_Post objects and ids in the same array?

Reply:
Unlikely. But aren't we trying to get all the post IDs that were found?

Reply:
There are two cases here. 

1. We have an array of post objects. Running array_map of get_post, will do nothing. 
2. We have an array of ids. We need to get an array of ids, pass it to `_prime_post_caches`, prime all the posts in a single request. Then array_map of get_post, will load from cache.

Reply:
My night brain had a similar thought (basically we don't need to re-prime the primed `WP_Post` objects) so I will give this a go.

Reply:
Ah, `fields: id=>parent` populates the `$posts` property with stdObj rather than WP_Posts, so I will need to account for that.